pages/blog/posts/understanding-lexical-dynamic-scopes.md

```diff
@@ -0,0 +1,244 @@
+---
+title: "Understanding JSON Schema Lexical and Dynamic Scopes"
+date: "2024-01-19"
+tags:
+  - Fundamentals
+  - Scopes
+type: Engineering
+cover: /img/posts/2024/understanding-lexical-dynamic-scopes/cover.webp
+authors:
+  - name: Juan Cruz Viotti
+    photo: /img/avatars/jviotti.webp
+    link: https://www.jviotti.com
+    byline: O'Reilly author, founder of Intelligence.AI and Sourcemeta
+excerpt: "A deep dive on lexical and dynamic scopes, essential for understanding more advanced topics like dynamic referencing"
+---
+
+Most of the keywords defined by the JSON Schema organization can be either
+evaluated on their own, or by only considering the values of their adjacent
+keywords. For example, the `type` keyword is independent of any other keywords,
+while the `minContains` and `maxContains` keywords depend on the `contains`
+keyword defined in the same schema object.
+
+However, there is a small set of keywords whose evaluation depend on the
+<i>scope</i> they are in. These keywords are `$ref`, `$dynamicRef`,
+`unevaluatedItems`, and `unevaluatedProperties`. Additionally, there is also a
+set of keywords that <i>affect</i> the scope they are declared in. These keywords
+are `$id`, `$schema`, `$anchor`, and `$dynamicAnchor`.
+
+JSON Schema introduces two types of scopes: the <i>lexical scope</i> and the
+<i>dynamic scope</i>. Understanding how these scopes work is essential for mastering
+some of the most advanced (and often confusing!) features of JSON Schema, such
+as dynamic referencing.
+
+Lexical Scope
+-------------
+
+JSON Schema is a circular data structure. What we mean by this is that as a
+schema author, you may use <i>applicator</i> keywords that take other schemas as
```

Comment:
Maybe worth crosslinking to the glossary for some of what's here, though there are lots of terms still missing, I think `applicator` isn't yet there...

Reply:
Yeah, good point. Let me cross link to anything that's already defined in there. I agree it'd be great to expand the list, so I can also cross link more from learnjsonschema.com